File: cli/src/auth/pkce.ts
/**
 * PKCE (Proof Key for Code Exchange) Utilities
 * Used for secure OAuth 2.0 authorization code flow
 */

import type { PKCEChallenge } from './types.js'

/**
 * Generate a cryptographically random code verifier
 * RFC 7636 specifies: 43-128 characters from [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"
 */
export function generateCodeVerifier(): string {
  const length = 43
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~'
  const randomBytes = crypto.getRandomValues(new Uint8Array(length))
  return Array.from(randomBytes)
    .map((b) => chars[b % chars.length])
    .join('')
}

/**
 * Generate a code challenge from a verifier using SHA-256
 * Returns base64url-encoded hash
 */
export async function generateCodeChallenge(verifier: string): Promise<string> {
  const encoder = new TextEncoder()
  const data = encoder.encode(verifier)
  const hash = await crypto.subtle.digest('SHA-256', data)
  return base64UrlEncode(hash)
}

/**
 * Generate both verifier and challenge for PKCE
 */
export async function generatePKCE(): Promise<PKCEChallenge> {
  const verifier = generateCodeVerifier()
  const challenge = await generateCodeChallenge(verifier)
  return { verifier, challenge }
}

/**
 * Generate a random state string for CSRF protection
 */
export function generateState(): string {
  const randomBytes = crypto.getRandomValues(new Uint8Array(32))
  return base64UrlEncode(randomBytes.buffer)
}

/**
 * Base64url encode an ArrayBuffer
 * RFC 4648 Section 5 - URL-safe base64 without padding
 */
function base64UrlEncode(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer)
  const binary = String.fromCharCode(...bytes)
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

File: crates/ava-auth/src/pkce.rs
//! PKCE (Proof Key for Code Exchange) generation.
//!
//! Generates cryptographically secure code verifier, challenge, and state parameters
//! per RFC 7636.

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use sha2::{Digest, Sha256};

/// PKCE parameters for an OAuth authorization request.
#[derive(Debug, Clone)]
pub struct PkceParams {
    /// Code verifier — 64 random bytes base64url-encoded (~86 chars).
    pub verifier: String,
    /// Code challenge — SHA-256(verifier) base64url-encoded.
    pub challenge: String,
    /// Random state parameter for CSRF protection — 32 random bytes base64url-encoded.
    pub state: String,
}

/// Generate cryptographically secure PKCE parameters.
pub fn generate_pkce() -> PkceParams {
    // Code verifier: 64 random bytes → base64url (no padding)
    let verifier_bytes: [u8; 64] = rand::random();
    let verifier = URL_SAFE_NO_PAD.encode(verifier_bytes);

    // Code challenge: SHA-256(verifier) → base64url (no padding)
    let digest = Sha256::digest(verifier.as_bytes());
    let challenge = URL_SAFE_NO_PAD.encode(digest);

    // State: 32 random bytes → base64url (no padding)
    let state_bytes: [u8; 32] = rand::random();
    let state = URL_SAFE_NO_PAD.encode(state_bytes);

    PkceParams {
        verifier,
        challenge,
        state,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pkce_generates_valid_lengths() {
        let pkce = generate_pkce();

        // Verifier: 64 bytes → 86 chars base64url
        assert_eq!(pkce.verifier.len(), 86);
        // Challenge: 32 bytes (SHA-256) → 43 chars base64url
        assert_eq!(pkce.challenge.len(), 43);
        // State: 32 bytes → 43 chars base64url
        assert_eq!(pkce.state.len(), 43);
    }

    #[test]
    fn pkce_challenge_matches_verifier() {
        let pkce = generate_pkce();

        // Recompute challenge from verifier
        let digest = Sha256::digest(pkce.verifier.as_bytes());
        let expected_challenge = URL_SAFE_NO_PAD.encode(digest);

        assert_eq!(pkce.challenge, expected_challenge);
    }

    #[test]
    fn pkce_generates_unique_values() {
        let a = generate_pkce();
        let b = generate_pkce();

        assert_ne!(a.verifier, b.verifier);
        assert_ne!(a.state, b.state);
    }

    #[test]
    fn pkce_uses_url_safe_chars() {
        let pkce = generate_pkce();
        let is_url_safe = |s: &str| {
            s.chars()
                .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
        };

        assert!(is_url_safe(&pkce.verifier));
        assert!(is_url_safe(&pkce.challenge));
        assert!(is_url_safe(&pkce.state));
    }
}
